include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,808 @@
+
+    /**
+     * Constructor which takes the FiniteElement @p projection_fe , the quadrature
+     * rule @p mass_quadrature used to integrate its local mass matrix and
+     * finally the quadrature rule @p data_quadrature which is used to store @p DataType .
+     */
+    ContinuousQuadratureDataTransfer(const FiniteElement<dim> &projection_fe,
+                                     const Quadrature<dim> &mass_quadrature,
+                                     const Quadrature<dim> &data_quadrature);
+
+    /**
+     * Virtual destructor.
+     */
+    ~ContinuousQuadratureDataTransfer();
+
+    /**
+     * Prepares for coarsening and refinement of a triangulation @p tria .
+     * @p data_storage represents the cell data which should be transfered and
+     * it should be initialized for each locally owned active cell.
+     *
+     * @note Although CellDataStorage class allows storing on different cells
+     * different objects derived from the base class, here we assume that
+     * @p data_storage contains objects of the same type, more specifically
+     * they pack/unpack the same data.
+     */
+    void prepare_for_coarsening_and_refinement (parallel::distributed::Triangulation<dim> &tria,
+                                                CellDataStorage<CellIteratorType,DataType> *data_storage);
+
+    /**
+     * Interpolate the data previously stored in this object before the mesh
+     * was refined or coarsened onto the quadrature points of the currently active
+     * set of cells.
+     *
+     * @note Before calling this function the user is expected to populate the
+     * data stored in the @p data_storage object provided to prepare_for_coarsening_and_refinement()
+     * at new cells using CellDataStorage::initialize(). If that is not the case,
+     * an exception will be thrown in debug mode.
+     */
+    void interpolate ();
+
+  private:
+    /**
+     * A callback function used to pack the data on the current mesh into
+     * objects that can later be retrieved after refinement, coarsening and
+     * repartitioning.
+     */
+    void pack_function(const typename parallel::distributed::Triangulation<dim,dim>::cell_iterator &cell,
+                       const typename parallel::distributed::Triangulation<dim,dim>::CellStatus status,
+                       void *data);
+
+    /**
+     * A callback function used to unpack the data on the current mesh that
+     * has been packed up previously on the mesh before refinement,
+     * coarsening and repartitioning.
+     */
+    void unpack_function (const typename parallel::distributed::Triangulation<dim,dim>::cell_iterator &cell,
+                          const typename parallel::distributed::Triangulation<dim,dim>::CellStatus status,
+                          const void *data);
+
+    /**
+     * FiniteElement used to project data from and to quadrature points.
+     */
+    const std::shared_ptr<const FiniteElement<dim> > projection_fe;
```

Comment:
Since you `clone()` the element, I think what you really want is a `unique_ptr`.